Client/Frontend/Browser/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,530 @@
+        } else {
+            completion(self.url, self.monogramFavicon)
+        }
+    }
+    
+    private static var attributesForHTMLCache: [URL: FaviconAttributes] = [:]
+    
+    /// Download the HTML of a page and use that to parse favicons out of the
+    /// `<head>` tags.
+    ///
+    /// - Note: Ensure any changes regarding icon fetching/scoring is also
+    /// updated in `MetadataHelper.js` to ensure both methods of fetching
+    /// favicons remain consistent.
+    private func parseHTMLForFavicons(for url: URL, _ completion: @escaping (FaviconAttributes) -> Void) {
+        let pageTask = session.dataTask(with: url) { [weak self] (data, response, error) in
+            guard let self = self else { return }
+            guard let data = data, error == nil, let root = try? HTMLDocument(data: data) else {
+                completion(self.monogramFavicon)
+                return
+            }
+            // Ensure page is reloaded to final landing page before looking for
+            // favicons
+            var reloadUrl: URL?
+            for meta in root.xpath("//head/meta") {
+                if let refresh = meta["http-equiv"], refresh == "Refresh",
+                    let content = meta["content"],
+                    let index = content.range(of: "URL="),
+                    let url = NSURL(string: String(content.suffix(from: index.upperBound))) {
+                    reloadUrl = url as URL
+                }
+            }
+            
+            if let url = reloadUrl {
+                self.parseHTMLForFavicons(for: url, completion)
+                return
+            }
+            
+            let xpath = self.kind == .largeIcon ?
+                "//head//link[contains(@rel, 'apple-touch-icon')]" :
+                "//head//link[contains(@rel, 'icon')]"
+            var highestScore: Double = -1.0
+            var icon: Favicon?
+            // Look for a favicon with the sizes closest to the goal
+            let goal = self.kind == .largeIcon ? 180.0 * 180.0 : 48 * 48
+            for link in root.xpath(xpath) {
+                guard let href = link["href"] else { continue }
+                let size = link["sizes"]?
+                    .split(separator: "x")
+                    .compactMap { Double($0) }
+                    .reduce(0, { $0 * $1 }) ?? 0.0
+                let score = 1.0 - (abs(size - goal)) / goal
+                
+                if score > highestScore, let faviconURL = URL(string: href, relativeTo: url) {
+                    highestScore = score
+                    icon = Favicon(url: faviconURL.absoluteString, date: Date(), type: self.kind.iconType)
+                }
+            }
+            
+            guard let favicon = icon, let faviconURL = URL(string: favicon.url) else {
+                completion(self.monogramFavicon)
+                return
+            }
+            
+            self.downloadIcon(url: faviconURL, addingToDatabase: true) { [weak self] image in
```

Comment:
Check for `isPrivate` before adding to database?

Reply:
It is already checked on completion on line 220: https://github.com/brave/brave-ios/pull/2490/files/6740bd8a934071e866f7db5d53b1d809c9da925c#diff-d47742238791692d201e5deaa57d0377R220

```
if addingToDatabase && !PrivateBrowsingManager.shared.isPrivateBrowsing {
  FaviconMO.add(favicon, forSiteUrl: self.url)
}
```